hawkular-inventory-api/src/main/java/org/hawkular/inventory/api/model/IdentityHash.java

```diff
@@ -0,0 +1,810 @@
+                            .slide(1, 0);
+
+                    return structure.getChildren(p, DataEntity.class)
+                            .filter(d -> returnType.equals(d.getRole()))
+                            .findFirst().orElse(dummyDataBlueprint(returnType));
+                }
+            };
+        }
+
+        default List<ResourceType.Blueprint> getResourceTypes() {
+            return Collections.emptyList();
+        }
+
+        default List<MetricType.Blueprint> getMetricTypes() {
+            return Collections.emptyList();
+        }
+
+        default List<OperationType.Blueprint> getOperationTypes(ResourceType.Blueprint rt) {
+            return Collections.emptyList();
+        }
+
+        default DataEntity.Blueprint<?> getReturnType(RelativePath rootResourceType, OperationType.Blueprint ot) {
+            return dummyDataBlueprint(returnType);
+        }
+
+        default DataEntity.Blueprint<?> getParameterTypes(RelativePath rootResourceType, OperationType.Blueprint ot) {
+            return dummyDataBlueprint(parameterTypes);
+        }
+
+        default DataEntity.Blueprint<?> getConfigurationSchema(ResourceType.Blueprint rt) {
+            return dummyDataBlueprint(configurationSchema);
+        }
+
+        default DataEntity.Blueprint<?> getConnectionConfigurationSchema(ResourceType.Blueprint rt) {
+            return dummyDataBlueprint(connectionConfigurationSchema);
+        }
+
+        default List<Resource.Blueprint> getResources(RelativePath rootPath, Resource.Blueprint parentResource) {
+            return Collections.emptyList();
+        }
+
+        default List<Resource.Blueprint> getFeedResources() {
+            return Collections.emptyList();
+        }
+
+        default List<Metric.Blueprint> getFeedMetrics() {
+            return Collections.emptyList();
+        }
+
+        default List<Metric.Blueprint> getResourceMetrics(RelativePath rootPath, Resource.Blueprint parentResource) {
+            return Collections.emptyList();
+        }
+
+        default DataEntity.Blueprint<?> getConfiguration(RelativePath rootPath, Resource.Blueprint parentResource) {
+            return dummyDataBlueprint(configuration);
+        }
+
+        default DataEntity.Blueprint<?> getConnectionConfiguration(RelativePath root, Resource.Blueprint
+                parentResource) {
+            return dummyDataBlueprint(connectionConfiguration);
+        }
+    }
+
+    private static class HashConstructor {
+        private final DigestComputingWriter digestor;
+
+        HashConstructor(DigestComputingWriter digestor) {
+            this.digestor = digestor;
+        }
+
+        public DigestComputingWriter getDigestor() {
+            return digestor;
+        }
+
+        public void startChild(IntermediateHashContext context) {
+//            System.out.println("start: " + context.root);
+        }
+
+        public void endChild(IntermediateHashContext ctx, IntermediateHashResult result) {
```

Comment:
I left the commented out debug statements there because I had so much trouble getting the computation right that I thought someone might use that again in the future ;)

This is a private class so I was not that fussed about purity but you're right - this could be abstract. It could even be made a member class of the `treeOf` method because that seems to be the only usage of it.